apps/heft/src/plugins/NodeServicePlugin.ts

```diff
@@ -0,0 +1,343 @@
+  private _nodeServiceConfiguration: INodeServicePluginConfiguration | undefined = undefined;
+  private _shellCommand: string | undefined;
+
+  // If true, then we will not attempt to relaunch the service until it is rebuilt
+  private _childProcessFailed: boolean = false;
+
+  private _pluginEnabled: boolean = false;
+
+  public apply(heftSession: HeftSession, heftConfiguration: HeftConfiguration): void {
+    this._logger = heftSession.requestScopedLogger('node-service');
+
+    this._isWindows = process.platform === 'win32';
+
+    heftSession.hooks.build.tap(PLUGIN_NAME, (build: IBuildStageContext) => {
+      build.hooks.loadStageConfiguration.tapPromise(PLUGIN_NAME, async () => {
+        this._nodeServiceConfiguration =
+          await CoreConfigFiles.nodeServiceConfigurationLoader.tryLoadConfigurationFileForProjectAsync(
+            this._logger.terminal,
+            heftConfiguration.buildFolder,
+            heftConfiguration.rigConfig
+          );
+
+        // defaults
+        this._configuration = {
+          commandName: 'serve',
+          ignoreMissingScript: false,
+          waitBeforeRestartMs: 2000,
+          waitForTerminateMs: 2000,
+          waitForKillMs: 2000
+        };
+
+        // TODO: @rushstack/heft-config-file should be able to read a *.defaults.json file
+        if (this._nodeServiceConfiguration) {
+          this._pluginEnabled = true;
+
+          if (this._nodeServiceConfiguration.commandName !== undefined) {
+            this._configuration.commandName = this._nodeServiceConfiguration.commandName;
+          }
+          if (this._nodeServiceConfiguration.ignoreMissingScript !== undefined) {
+            this._configuration.ignoreMissingScript = this._nodeServiceConfiguration.ignoreMissingScript;
+          }
+          if (this._nodeServiceConfiguration.waitBeforeRestartMs !== undefined) {
+            this._configuration.waitBeforeRestartMs = this._nodeServiceConfiguration.waitBeforeRestartMs;
+          }
+          if (this._nodeServiceConfiguration.waitForTerminateMs !== undefined) {
+            this._configuration.waitForTerminateMs = this._nodeServiceConfiguration.waitForTerminateMs;
+          }
+          if (this._nodeServiceConfiguration.waitForKillMs !== undefined) {
+            this._configuration.waitForKillMs = this._nodeServiceConfiguration.waitForKillMs;
+          }
+
+          this._shellCommand = (heftConfiguration.projectPackageJson.scripts || {})[
+            this._configuration.commandName
+          ];
+
+          if (this._shellCommand === undefined) {
+            if (this._configuration.ignoreMissingScript) {
+              this._logger.terminal.writeLine(
+                `The plugin is disabled because the project's package.json` +
+                  ` does not have a "${this._configuration.commandName}" script`
+              );
+            } else {
+              this._logger.terminal.writeErrorLine(
```

Comment:
Do you want `this._logger.emitError` instead?

Reply:
I changed it to `throw Error()` because none of those other options will abort the Heft operation